internal/wasm/memory.go

```diff
@@ -56,19 +57,63 @@ type MemoryInstance struct {
 	// waiters implements atomic wait and notify. It is implemented similarly to golang.org/x/sync/semaphore,
 	// with a fixed weight of 1 and no spurious notifications.
 	waiters sync.Map
+
+	mmappedMemory bool
+	closed        bool
 }
 
 // NewMemoryInstance creates a new instance based on the parameters in the SectionIDMemory.
-func NewMemoryInstance(memSec *Memory) *MemoryInstance {
+func NewMemoryInstance(memSec *Memory, compilerEngine bool) *MemoryInstance {
 	min := MemoryPagesToBytesNum(memSec.Min)
 	capacity := MemoryPagesToBytesNum(memSec.Cap)
+
+	var buffer []byte
+	var cap uint32
+	mmappedMemory := false
+	if memSec.IsShared && compilerEngine {
+		// When using compiler, memory accesses can happen at the same time that memory is resized, meaning
+		// we cannot have the memory base move during operation. mmap allows allocating memory virtually so
+		// we can grow without changing the base. The spec requires max for shared memory currently because
+		// all threads implementations are effectively expected to use mmap for shared memory.
```

Comment:
Is it really the only issue for compiler? I think interpreter can also have the same problem e.g. what happens if memory grow happens right before https://github.com/tetratelabs/wazero/blob/bfd9a00cd64e4380d5ab308fc85f9650a6370d9d/internal/engine/interpreter/interpreter.go#L1624 ?

Reply:
Ah I forgot when I first wrote this code, Grow was guarded by the same mutex but currently it's not. With that lock there could be no accessed I think.

Since interpreter can be used on platforms without mmap, should I restore the lock on grow, only when mmap isn't used? Let me know if it makes sense

Reply:
even grow was guarded,  accessing `memoryInst.Buffer` can be unsafe as e.g., what if grow happens right after dereferencing []byte? loads/stores are not guarded by mutex in the first place

Reply:
Ah that's true, it's not thread-safe code but also shouldn't crash. I guess it means we need to use mmap with interpreter too. I'll remove the checks in this code and in the future we can return an error is user specifies threads flag and mmap isn't supported?

Reply:
I think we simply use `make([]byte, maxpages)` always for interpreter, wdyt? cc @ncruces

Reply:
maxpages will often be something like `4GB` since it's assuming use of mmap. It's because it's hard to predict the actual required max memory when compiling a binary since it can be run in many different ways.

https://github.com/wasix-org/wasix-libc/blob/main/.github/workflows/files/clang-wasix.cmake_toolchain#L5
https://github.com/wasilibs/go-re2/blob/main/buildtools/re2/Dockerfile#L51

Reply:
I think I'm fine with it until someone screams assuming few people will use the thread proposal on interpreter env

Reply:
Ok I went ahead and went with that. One use case I heard of interpreter was compiling sqlc itself to wasm to run. But maybe we can tweak later

Reply:
I agree this seems the only sane way to do it. Users that scream have remedy: edit their WASM to reduce max memory.

Does this mean we can avoid `IsCompiler`, use `mmap` for shared where we can, huge slice where we can't?

Reply:
Yeah that's also an option, want to confirm it's OK to use mmap with interpreter when it's supported as I believe so far we haven't used much syscall in that codepath.